### PR TITLE
Refactor register.c to remove calls to STRLEN()

### PR DESCRIPTION
**Pull request by basilisk0315**
Added new struct string_T to store a string with it's length.
Also, fix a bug in charwise_block_prep() in ops.c where bdp->textlen is off-by-one if the selection runs over the end of the line. Adjust test_visual.vim accordingly.

### CODE REVIEW

**Comment by zeertzjq** [comment]
> Adjust test_visual.vim accordingly.

Why? Shouldn't `f_getregionpos()` be fixed instead?

**Comment by basilisk0315** [comment]
f_getregionpos() calls charwise_block_prep() (in ops.c) to retrieve starting positions and offsets which is then reported. This function has the off-by-one error (eg string 'one' was being reported at having a ->textlen of 4, instead of 3 if it occurs at the end of a line). test_visual.vim was checking for the higher number.

Have I taken the wrong approach here?

**Comment by zeertzjq** [comment]
This makes the ending column of charwise selection inconsistent with linewise selection. With `eol: v:false`, the ending column is one more than the line length for every line in a linewise selection and every non-last line in the charwise selection, but with PR it is no longer than case for charwise selection.

Also, with `eol: v:false`,
> if a line is empty or the selection is entirely beyond the end of a line, a "col" value of 0 is used for both positions.

This can happen for charwise selection with 'virtualedit', as can be seen in this test case:
```vim
    call cursor(1, 3)
    call feedkeys("\<ESC>lvjj3l", 'tx')
    call assert_equal(['', 'two', 'three  '],
          \ getregion(getpos('v'), getpos('.'), {'type': 'v' }))
    call assert_equal([
          \   [[bufnr('%'), 1, 0, 0], [bufnr('%'), 1, 0, 0]],
          \   [[bufnr('%'), 2, 1, 0], [bufnr('%'), 2, 3, 0]],
          \   [[bufnr('%'), 3, 1, 0], [bufnr('%'), 3, 5, 0]],
          \ ],
          \ getregionpos(getpos('v'), getpos('.'), {'type': 'v' }))
```
With this PR the ending column changes to the length of the line, which is inconsistent with a blockwise selection entirely beyond the end of the line, for which the ending column is still 0.

**Comment by chrisbra** [comment]
Thanks. Can you back out the change to `getregion()`? So we have a single PR for one single fix/enhancement?

**Comment by basilisk0315** [comment]
I will back out the change to ops.c and the test. Stay tuned...

File: src/register.c [hunk 1]
```diff
@@ -1153,8 +1166,8 @@ op_yank(oparg_T *oap, int deleting, int mess)
 #if defined(FEAT_CLIPBOARD) && defined(FEAT_X11)
     int			did_star = FALSE;
 #endif
+    int			i;
 
-				    // check for read-only register
```
Comment (zeertzjq) [inline src/register.c]: Why is this comment removed? A similar comment still exists before `valid_yank_reg()` in `stuff_yank()`.

File: src/register.c [hunk 2]
```diff
@@ -1198,12 +1211,14 @@ op_yank(oparg_T *oap, int deleting, int mess)
...
     y_current->y_width = 0;
-    y_current->y_array = lalloc_clear(sizeof(char_u *) * yanklines, TRUE);
+    y_current->y_array = lalloc_clear(sizeof(string_T) * yanklines, TRUE);
     if (y_current->y_array == NULL)
     {
 	y_current = curr;
 	return FAIL;
     }
+    for (i = 0; i < y_current->y_size; ++i)
+	y_current->y_array[i].length = 0;
```
Comment (zeertzjq) [inline src/register.c]: Is this needed? Shouldn't `lalloc_clear()` have already set them to zero?

File: src/register.c [hunk 3]
```diff
@@ -2116,24 +2181,25 @@ do_put(
...
+
 			curwin->w_cursor.lnum = lnum;
 			ptr = ml_get(lnum);
 			if (cnt == count && i == y_size - 1)
 			    lendiff = ml_get_len(lnum);
 			if (*ptr == '#' && preprocs_left())
 			    indent = 0;     // Leave # lines at start
 			else
-			     if (*ptr == NUL)
+			if (*ptr == NUL)
```
Comment (zeertzjq) [inline src/register.c]: Shouldn't this `if` be joined with the previous `else`?

File: src/register.c [hunk 4]
```diff
@@ -1232,14 +1247,27 @@ op_yank(oparg_T *oap, int deleting, int mess)
...
 	    case MCHAR:
 		{
+		    int  tmp;
+
 		    charwise_block_prep(oap->start, oap->end, &bd, lnum, oap->inclusive);
+
+		    // make sure bd.textlen is not longer than the text
+		    tmp = (int)STRLEN(bd.textstart);
+		    if (tmp < bd.textlen)
+			bd.textlen = tmp;
```
Comment (chrisbra) [inline src/register.c:1263]: shouldn't that part belong into `charwise_block_prep()`?

File: src/register.c [hunk 5]
```diff
@@ -1431,10 +1467,13 @@ yank_copy_line(struct block_def *bd, long y_idx, int exclude_trailing_space)
 	while (s > 0 && VIM_ISWHITE(*(bd->textstart + s - 1)))
 	{
 	    s = s - (*mb_head_off)(bd->textstart, bd->textstart + s - 1) - 1;
-	    pnew--;
+	    --pnew;
```
Comment (chrisbra) [inline src/register.c]: unrelated style change

File: src/register.c [hunk 6]
```diff
@@ -2073,7 +2128,17 @@ do_put(
 	else
 	{
 	    linenr_T	new_lnum = new_cursor.lnum;
-	    size_t	len;
+	    int		indent;
+	    int		orig_indent;
```
Comment (chrisbra) [inline src/register.c]:
```suggestion
	    int		orig_indent = 0;
```

and then get rid of the lines below:
```
    else
		orig_indent = 0;
```

File: src/register.c [hunk 7]
```diff
@@ -2116,24 +2181,25 @@ do_put(
 		{
 		    if (y_type != MCHAR || i < y_size - 1)
 		    {
-			if (ml_append(lnum, y_array[i], (colnr_T)0, FALSE)
+			if (ml_append(lnum, y_array[i].string, (colnr_T)0, FALSE)
 								      == FAIL)
 			    goto error;
-			new_lnum++;
+			++new_lnum;
```
Comment (chrisbra) [inline src/register.c]: unrelated style change

File: src/register.c [hunk 8]
```diff
@@ -2116,24 +2181,25 @@ do_put(
...
 		    }
-		    lnum++;
+		    ++lnum;
```
Comment (chrisbra) [inline src/register.c]: unrelated style change

File: src/register.c [hunk 9]
```diff
@@ -260,6 +259,7 @@ get_yank_register(int regname, int writing)
     y_current = &(y_regs[i]);
     if (writing)	// remember the register we write into for do_put()
 	y_previous = y_current;
+
```
Comment (chrisbra) [inline src/register.c]: unrelated new line

File: src/register.c [hunk 10]
```diff
@@ -2650,9 +2715,10 @@ get_reg_contents(int regname, int flags)
     len = 0;
     for (i = 0; i < y_current->y_size; ++i)
     {
-	len += (long)STRLEN(y_current->y_array[i]);
-	// Insert a newline between lines and after last line if
-	// y_type is MLINE.
+	len += (long)y_current->y_array[i].length;
+
+	// Insert a NL between lines and after the last line if y_type is
```
Comment (chrisbra) [inline src/register.c]: NL -> newline please

**Review by chrisbra — COMMENTED** [review]
Thanks. I noticed there are now a few:
```
    vim_free(y_current->y_array[0].string);
     y_current->y_array[0].length = 0;
```
I think it may make sense to have a new `vim_free_clear_len(string_T)` function,
that does this once.

File: src/register.c (hunk 2 above)
Comment (basilisk0315) [reply src/register.c]: Good pickup thanks

File: src/register.c (hunk 1 above)
Comment (basilisk0315) [reply src/register.c]: I got a bit trigger happy with 'dd'.

File: src/register.c (hunk 3 above)
Comment (basilisk0315) [reply src/register.c]: Yep

File: src/register.c (hunk 4 above)
Comment (zeertzjq) [reply src/register.c:1263]: That'll change the behavior of `getregionpos()`

Comment (basilisk0315) [reply src/register.c:1263]: Putting this into charwise_block_prep() bring us back to the consistency issue as pointed out by @zeertzjq above.

File: src/alloc.c [hunk 11]
```diff
@@ -617,6 +617,13 @@ vim_free(void *x)
     }
 }
 
+    void
+vim_free_string(string_T *s)
+{
+    VIM_CLEAR(s->string);
+    s->length = 0;
+}
```
Comment (zeertzjq) [inline src/alloc.c]: Maybe this can be changed to a macro instead?
```c
#define VIM_CLEAR_STRING(s) \
    do { \
	VIM_CLEAR(s.string); \
	s.length = 0; \
    } while (0)
```

Comment (basilisk0315) [reply src/alloc.c]: Ok.

**Comment by chrisbra** [comment]
can you please re-run? I have seen this sometimes, but not sure why the xml indent test fails

File: src/register.c (hunk 4 above)
Comment (chrisbra) [reply src/register.c:1263]: then we should back this change out, no? Or does this fix an actual bug?

File: src/viminfo.c [hunk 12]
```diff
@@ -1726,21 +1730,27 @@ read_viminfo_register(vir_T *virp, int force)
...
 	    // Move the lines from array[] to y_array[].
-	    y_current_p->y_array = ALLOC_MULT(char_u *, size);
+	    y_current_p->y_array = ALLOC_MULT(string_T, size);
 	    for (i = 0; i < size; i++)
 	    {
 		if (y_current_p->y_array == NULL)
-		    vim_free(array[i]);
+		{
+		    vim_free(array[i].string);
+		    array[i].length = 0;
```
Comment (chrisbra) [inline src/viminfo.c]:
```suggestion
		   VIM_CLEAR_STRING(array[i]);
```

File: src/register.c (hunk 4 above)
Comment (basilisk0315) [reply src/register.c:1263]: This piece of code makes sure that the string length of the register (specifically) is correct, while maintaining consistent behaviour in other parts of the code that use charwise_block_prep().

**Comment by chrisbra** [comment]
thanks!
